[Maxx Chan Yuhang] iP

maxxyh opened the pull request.

Commit:
Add Gradle support

maxxyh committed:
Level-1. Greet, Echo, Exit

- A skeletal version of Duke that starts by greeting the user, simply echos commands entered by the user, and exits when the user types bye.
- Changed name to Alpha

maxxyh committed:
Level 2. Add, List

Add the ability to store whatever text entered by the user and display them back to the user when requested.

maxxyh committed:
Level 3. Mark as Done

- Add the ability to mark tasks as done. 
- Also added a check to make sure cannot set entries which don't exist to be done.

maxxyh committed:
Level 4. ToDos, Events, Deadlines

Add support for tracking three types of tasks:
1. ToDos: tasks without any date/time attached to it e.g., visit new theme park
2. Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
3. Events: tasks that start at a specific time and ends at a specific time e.g., team project meeting on 2/10/2019 2-4pm

maxxyh committed:
A-TextUiTesting: Automated Text UI Testing

Use the input/output redirection technique to semi-automate the testing of Duke.

maxxyh committed:
Level 5. Handle Errors

Teach Alpha to deal with errors such as incorrect inputs entered by the user.
Alpha now handles the following errors: 
1. no task specified for tasks
2. done command with an invalid index/ no index
3. invalid command

maxxyh committed:
Minor Testing Update to Level 5. Handle Errors

Updated UI testing suite for level 5

maxxyh committed:
Level 6. Delete

Added support for deleting tasks from the list.
Also implemented DukeException for the delete command, similar to the done command.
Also updated the testing suite.

maxxyh committed:
A-Enums

Use enums for commands instead of string comparison

Ben-Hanan commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,123 @@
-                + "| | | | | | | |/ / _ \\\n"
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
+        String logo = "   ___      _      _ __   _              \n" +
+                "  /   \\    | |    | '_ \\ | |_     __ _   \n" +
+                "  | - |    | |    | .__/ | ' \\   / _` |  \n" +
+                "  |_|_|   _|_|_   |_|__  |_||_|  \\__,_|  \n" +
+                "_|\"\"\"\"\"|_|\"\"\"\"\"|_|\"\"\"\"\"|_|\"\"\"\"\"|_|\"\"\"\"\"| \n" +
+                "\"`-0-0-'\"`-0-0-'\"`-0-0-'\"`-0-0-'\"`-0-0-' ";
         System.out.println("Hello from\n" + logo);
+
+        String greeting = "    ____________________________________________________________\n" +
+                "     Hello, Alpha here... welcome to my help centre... again :/\n" +
+                "     Would you like to explain what you want?\n" +
+                "    ____________________________________________________________";
```
Since the logo and the greeting is constant maybe they should be static instead? And since they are static then the names should adhere to the naming conventions!

Ben-Hanan commented on src/main/java/EventTask.java:
```diff
@@ -0,0 +1,18 @@
+    @Override
+    public String toString()
+    {
+        return String.format("[E][%s] %s (at: %s)", getStatusIcon(), description, eventTime);
```
Maybe the type of the event represented by the "E" here should be static since it's a constant?

Ben-Hanan commented on src/main/java/ToDoTask.java:
```diff
@@ -0,0 +1,12 @@
+    @Override
+    public String toString()
+    {
+        return String.format("[T][%s] %s", getStatusIcon(), description);
```
Maybe the type of the event represented by the "T" here should be static since it's a constant?

Ben-Hanan reviewed (commented):
LGTM! This is a nice implementation of Duke up to level of A-Enums. Not sure where the rest of the project is but I'm sure it'll be of similar quality as well.

maxxyh committed:
Level-7

Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when Duke starts up. You may hard-code the file name and location

maxxyh committed:
Level-8: Dates and Times

Deadlines and Event times now stored as DateTime objects

maxxyh committed:
Merge branch 'branch-Level-7'

maxxyh committed:
Merge branch 'branch-Level-8'

Marcon2509 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,224 @@
+        try {
+            // read the file
+            List<String> dataLines = Files.readAllLines(path);
+            return DecodeTasksFromSave(dataLines);
+        // should be handled above so it's an assertion instead of an exception
+        } catch (FileNotFoundException fnfe) {
+            throw new AssertionError("A non-existent file scenario should already have been handled earlier.");
+        // io exception for example no permission to read a file, some process closed the stream
+        } catch (IOException ioe) {
+            throw new StorageOperationException("Error reading file: " + path);
+        // illegalValueException is when arguments cannot form a task
+        } catch (IllegalValueException ive) {
+            throw new StorageOperationException("File contains illegal data values; data type constraints not met");
+        }
+    }
+
+    private static List<Task> DecodeTasksFromSave(List<String> dataLines) throws IllegalValueException {
+        List<Task> output = new ArrayList<>();
+        for (String line : dataLines) {
+            Matcher matcher = TASK_DATA_ARGS_FORMAT.matcher(line);
+            if (!matcher.matches()) {
+                throw new IllegalValueException("Encoded task in invalid format. Unable to decode.");
+            }
+            String command = matcher.group("eventType");
+            boolean isDone = matcher.group("isDone").equals("0") ? false : true;
+            String description  = matcher.group("description");
+            if (command.equals("T")) {
+                output.add(new ToDoTask(description, isDone));
+            } else if (command.equals("E")) {
+                output.add(new EventTask(description, isDone, matcher.group("date")));
+            } else if (command.equals("D")) {
+                output.add(new DeadlineTask(description, isDone, matcher.group("date")));
+            } else {
+                throw new IllegalValueException("Encoded task in invalid format. Unable to decode.");
+            }
+        }
+
+        return output;
+    }
+
```
Should there be javaDoc comments for the methods?

Marcon2509 commented on src/main/java/EventTask.java:
```diff
@@ -0,0 +1,47 @@
+        endTime = LocalTime.parse(endTemp.substring(0,2) + ":"
+                + endTemp.substring(2,4));
+    }
+
+    public EventTask(String description, boolean done, String eventTime) {
+        super(description);
+        isDone = done;
+        this.eventTime = eventTime;
+    }
+
+    @Override
+    public String[] getSaveData() {
+        return new String[] {"E", isDone ? "1" : "0", description, eventTime};
+    }
+
+    @Override
+    public String toString()
+    {
+        return String.format("[E][%s] %s (at: %s)", getStatusIcon(), description,
+                String.format(
+                        eventDate.format(DateTimeFormatter.ofPattern("MMM d yyyy"))
+                        + ", " + startTime.format(DateTimeFormatter.ofPattern("Hmm'hrs'"))
+                        + "-" + endTime.format(DateTimeFormatter.ofPattern("Hmm'hrs'"))
+                        ));
+    }
+
+}
```
Should there be javadoc comments?

Marcon2509 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,224 @@
+import jdk.jfr.Event;
+
+import java.io.*;
```
Should import java.io.* be used?

Marcon2509 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,224 @@
+    }
+
+    private static void printAddTaskListStatus(List<Task> taskList)
+    {
+        int size = taskList.size();
+        System.out.println("    Fine. I've added this task:") ;
+        System.out.println("      " + taskList.get(size - 1));
+        System.out.println("    Now you have a total of " + taskList.size() + (taskList.size() == 1 ? " task" : " tasks") +  " in your list.");
+    }
+
```
Perhaps use a method that helps you indent all System.out.print lines

Marcon2509 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,224 @@
+            List<String> dataLines = Files.readAllLines(path);
+            return DecodeTasksFromSave(dataLines);
+        // should be handled above so it's an assertion instead of an exception
+        } catch (FileNotFoundException fnfe) {
```
perhaps use a better name than "fnfe"

Marcon2509 reviewed (commented):
Overall pretty good although im not sure where your classes needed for the GUI are but im sure it will be just as good in quality

maxxyh committed:
A-MoreOOP

Made solution more OOP with the addition of the Ui,
Storage, Parser and TaskList classes.

maxxyh committed:
A-MoreOOP Part 2

Uncommitted files from last commit

maxxyh committed:
A-Packages

Organised classes into 5 sub-packages: 
1. commands
2. exceptions
3. storage
4. tasks
5. ui

maxxyh committed:
A-JUnit

Added 3 unit tests: 1 testing "bye" termination in duke.java, and 2 testing the parser for event and deadline task inputs.

maxxyh committed:
Minor bug fixes for Duke Constructor

maxxyh committed:
A-JavaDoc

Added JavaDoc comments to major classes,
public methods, and significant non-public methods.

maxxyh committed:
A-CodingStandard

Cleaned up syntax for code, 
removed unnecessary imports, 
added new "undone task count" functionality 
to "list" command

maxxyh committed:
Level-9. Find

Added a "find" function for users to find a task by 
searching for a keyword (non-case sensitive).

maxxyh committed:
Merge branch 'branch-A-JavaDoc'

maxxyh committed:
Merge branch 'branch-A-CodingStandard'

maxxyh committed:
Merge branch 'branch-Level-9'

maxxyh committed:
Merge branch 'add-gradle-support'

maxxyh committed:
Minor tweaks for Gradle

maxxyh committed:
A-CheckStyle

Set up CheckStyle in Gradle and tweaked code to fit check style checks

maxxyh committed:
Merge branch 'branch-A-CheckStyle'

maxxyh committed:
Added GUI

Using FXML and JavaFX, Alpha now operates on a GUI.

maxxyh committed:
Merge branch 'branch-Level-10'

maxxyh committed:
Merge branch 'branch-Level-10'

Varargs is used in Ui outputBlockToUser method

maxxyh committed:
Merge remote-tracking branch 'origin/master'

maxxyh committed:
Added assertions

Documenting assumptions made at different points in the code.

maxxyh committed:
Refactor relationship between MainWindow Duke and Ui

maxxyh committed:
Merge pull request #1 from maxxyh/branch-A-Assertions

Added assertions

maxxyh committed:
Merge branch 'master' into A-CodeQuality

maxxyh committed:
Merge pull request #2 from maxxyh/A-CodeQuality

Refactor relationship between MainWindow Duke and Ui

maxxyh committed:
Postpone and Advance

Added commands to allow postponement and 
advancement of events and deadlines in the format:
(postpone/advance) (taskIndex) /by (amount) (time unit)

maxxyh committed:
Merge branch 'branch-BCD-Extension'

maxxyh committed:
Bug fix for advance

Fixed variable assignment bug in advance method 
calls, and made postpone() and advance() specific to 
Event and Deadline only.

maxxyh committed:
GUI Improvements

- Changed DialogBox to use Text to support overflow 
- DialogBox now has rounded corners, and has
 different background colours for user and duke
- added padding to DialogBox for visual spacing 
- Flipped Alpha's icon
- Errors will be displayed in red

maxxyh committed:
Draft of User Guide

Added all feature write-ups, but command summary in progress.

maxxyh committed:
Updated User Guide, Bug Fix

Fixed wrong index bug for done and delete. 
Updated user guide with product screenshots
Updated user guide with command summary

maxxyh committed:
Fixed table formatting for command summary

There should be space between the hyphens

maxxyh committed:
Refactoring: Commands, Storage, Testing, jar

Changed Command's execute to return a String to support testing and decouple it from Ui.
Changed save location to be consistently at the user home directory, and handled case where directory does not exist.
Added help function, which shows list of commands that can be called
Added icon to application.
Added new test cases for Parser and Commands
Edited Gradle settings to correctly generate jar file.

maxxyh committed:
Rename ui.png to Ui.png
